scipy/integrate/_quadrature.py

```diff
@@ -1051,65 +1045,82 @@ def cumulative_simpson(y, *, x=None, dx=1.0, axis=-1, initial=None):
     >>> ax.grid()
     >>> plt.show()
 
+    The output of `cumulative_simpson` is similar to that of iteratively
+    calling `simpson` with successively larger integration intervals, but
+    not identical.
+
+    >>> def cumulative_simpson_reference(y, x):
+    ...     return np.asarray([integrate.simpson(y[:i], x=x[:i])
+    ...                        for i in range(2, len(y) + 1)])
+    >>>
+    >>> rng = np.random.default_rng(354673834679465)
+    >>> x, y = rng.random(size=(2, 10))
+    >>> x.sort()
+    >>>
+    >>> res = integrate.cumulative_simpson(y, x=x)
+    >>> ref = cumulative_simpson_reference(y, x)
+    >>> equal = np.abs(res - ref) < 1e-15
+    >>> equal  # not equal when `simpson` has even number of subintervals
+    array([False,  True, False,  True, False,  True, False,  True,  True])
+
+    This is expected: because `cumulative_simpson` has access to more
+    information than `simpson`, it can typically produce more accurate
+    estimates of the underlying integral over subintervals.
+
     """
-    y = _convert_input_to_float_array(y)
+    y = _ensure_float_array(y)
 
-    # validate y and axis
-    if axis < -1 or axis >= y.ndim:
```

Comment:
This was a bit too restrictive in the range of `axis` values and not restrictive enough about the data type of `axis` (e.g. if `axis=1.5` there would be a weird error). This let's NumPy do the detailed input validation of `axis` and takes care of standardizing `y` so integration is always performed along the last axis.